WalletWasabi/BestEffortEndpointConnector.cs

```diff
@@ -0,0 +1,128 @@
+using NBitcoin.Protocol;
+using NBitcoin.Protocol.Behaviors;
+using NBitcoin.Protocol.Connectors;
+using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi
+{
+	public class BestEffortEndpointConnector : IEnpointConnector
+	{
+		public BestEffortEndpointConnector()
+			: this(
+				new DefaultEndpointConnector(allowOnlyTorEndpoints: true),
+				new EffortState(ConnectionMode.OnionServiceOnly, DateTimeOffset.UtcNow))
+		{
+		}
+
+		private BestEffortEndpointConnector(DefaultEndpointConnector connector, EffortState state)
+		{
+			Connector = connector;
+			State = state;
+		}
+
+		public enum ConnectionMode
+		{
+			// Connect to nodes published as onion services only
+			OnionServiceOnly,
+
+			// Connect to nodes published on the clearnet through Tor exit nodes
+			AllowGoingThroughTorExitNodes,
+		}
+
+		public DefaultEndpointConnector Connector { get; private set; }
+		public EffortState State { get; private set; }
+
+		public void UpdateConnectedNodesCounter(int connectedNodes)
+		{
+			State.ConnectedNodesCount = connectedNodes;
+		}
+
+		public IEnpointConnector Clone()
+		{
+			return new BestEffortEndpointConnector(Connector, State);
+		}
+
+		public virtual async Task ConnectSocket(Socket socket, EndPoint endPoint, NodeConnectionParameters nodeConnectionParameters, CancellationToken cancellationToken)
+		{
+			if (State.CheckModeUpdate())
+			{
+				Connector.AllowOnlyTorEndpoints = State.Mode switch
+				{
+					ConnectionMode.OnionServiceOnly => true,
+					ConnectionMode.AllowGoingThroughTorExitNodes => false,
+					_ => throw new InvalidOperationException($"Unknown ${typeof(ConnectionMode).Name} with value {State.Mode}.")
+				};
+			}
+
+			await Connector.ConnectSocket(socket, endPoint, nodeConnectionParameters, cancellationToken).ConfigureAwait(false);
+			State.LastModeChangeTime = DateTimeOffset.UtcNow;
```

Comment:
I'm not sure why connecting a peer is considered to be a mode change. Can you explain?

Reply:
I have to change this.